### PR TITLE
eip7251: Bugfix and more withdrawal tests

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**
implements https://github.com/ethereum/consensus-specs/pull/3979 for devnet 5

should be updated along with new spec tests.

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have made an appropriate entry to [CHANGELOG.md](https://github.com/prysmaticlabs/prysm/blob/develop/CHANGELOG.md).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

### CODE REVIEW

File: beacon-chain/core/blocks/withdrawals.go [hunk 1]
```diff
@@ -120,35 +120,36 @@ func ValidateBLSToExecutionChange(st state.ReadOnlyBeaconState, signed *ethpb.Si
 //
 // Spec pseudocode definition:
 //
-//	def process_withdrawals(state: BeaconState, payload: ExecutionPayload) -> None:
-//	    expected_withdrawals, partial_withdrawals_count = get_expected_withdrawals(state)  # [Modified in Electra:EIP7251]
+//		def process_withdrawals(state: BeaconState, payload: ExecutionPayload) -> None:
+//		    # [Modified in Electra:EIP7251]
+//	   	expected_withdrawals, processed_partial_withdrawals_count = get_expected_withdrawals(state)
```
Comment (terencechain) [inline beacon-chain/core/blocks/withdrawals.go:125]: Can we revert these changes? I think the original format is fine?
Reason being it's hard to view the differences and the next person will just change it back agian

File: CHANGELOG.md [hunk 2]
```diff
@@ -43,6 +43,7 @@ The format is based on Keep a Changelog, and this project adheres to Semantic Ve
 - Fixed mesh size by appending `gParams.Dhi = gossipSubDhi`
 - Fix skipping partial withdrawals count.
 - wait for the async StreamEvent writer to exit before leaving the http handler, avoiding race condition panics [pr](https://github.com/prysmaticlabs/prysm/pull/14557)
+- EIP7521 - fixing withdrawal bug
```
Comment (terencechain) [inline CHANGELOG.md:46]: Can we be more specific with this change log? High level: what is the bug?

File: beacon-chain/state/state-native/getters_withdrawal.go [hunk 3]
```diff
@@ -155,6 +158,15 @@ func (b *BeaconState) ExpectedWithdrawals() ([]*enginev1.Withdrawal, uint64, err
 		if err != nil {
 			return nil, 0, errors.Wrapf(err, "could not retrieve balance at index %d", validatorIndex)
 		}
+		if b.version >= version.Electra {
+			var partiallyWithdrawnBalance uint64
+			for _, w := range withdrawals {
```
Comment (terencechain) [inline beacon-chain/state/state-native/getters_withdrawal.go:163]: I think this is fine and not really worth optimizing, it's nice to read along the spec, but if you want to optimize it, we could use a map or track the length of withdrawals before entering the for loop and break at that index.

File: CHANGELOG.md [hunk 4]
```diff
@@ -43,6 +43,7 @@ The format is based on Keep a Changelog, and this project adheres to Semantic Ve
 - Fixed mesh size by appending `gParams.Dhi = gossipSubDhi`
 - Fix skipping partial withdrawals count.
 - wait for the async StreamEvent writer to exit before leaving the http handler, avoiding race condition panics [pr](https://github.com/prysmaticlabs/prysm/pull/14557)
+- EIP7521 - fixing withdrawal bug, pending partial withdrawals was not taken account of in the withdrawal sweep. The already withdrawn balance should be deducted from the balance considered in the sweep.
```
Comment (saolyn) [inline CHANGELOG.md:46]: Sentence nit
```suggestion
- EIP7521 - Fixes withdrawal bug by accounting for pending partial withdrawals and deducting already withdrawn amounts from the sweep balance.
```

**Review by terencechain — APPROVED** [review]

**Review by terencechain — APPROVED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (coldpak) [reply CHANGELOG.md]: It's a really minor detail, but shouldn't it be 7251 instead of 7521?
